File: julia8024/BOJ_2512.py
# 백준
# 실버2 예산

import sys

N = int(sys.stdin.readline()) # 지방의 수
arr = list(map(int, sys.stdin.readline().split())) # 각 지방의 예산 요청
M = int(sys.stdin.readline()) # 총 예산

result = 0
start, end = 1, max(arr)


while start <= end:
    mid = (start + end) // 2 # 중앙값 계산
    total = 0 # 예산 합
    for i in arr:
        if i > mid:
            total += mid
        else:
            total += i
    
    if total <= M:
        # 중앙값+1 ~ 끝값 다시 찾기
        result = mid
        start = mid + 1
    else:
        # 시작 ~ 중앙값-1 다시 찾기
        end = mid - 1

print(result)

File: chaeryeon823/BOJ_2512.js
const fs = require("fs");
const input = fs.readFileSync(0, "utf8").trim().split("\n");

const N = Number(input[0]); // 지방 수
const req = input[1].split(" ").map(Number); // 각 지방의 예산 요청
const M = Number(input[2]); // 총 예산

let low = 0;
let high = Math.max(...req);
let answer = 0;

while (low <= high) {
  const mid = Math.floor((low + high) / 2);
  let total = 0;

  for (let i = 0; i < N; i++) {
    total += req[i] > mid ? mid : req[i];
  }

  if (total <= M) {
    answer = mid;
    low = mid + 1;
  } else {
    high = mid - 1;
  }
}

console.log(answer);
